### PR TITLE
Add ability to match/assert on all the args at once

### LINKED ISSUE #66: Feature request: allow partial argument matches
## overview

I'm a big fan of jest-when, and it makes using jest mocks _far_ more pleasant than using them vanilla. I have a feature request (that I'd be very open to PR'ing, if you find it acceptable) to cover a particular usage that we've run into at my company.

Currently, `calledWith` requires all function arguments to be specified. For the vast majority of use cases, I think this is sensible and correct. I think there are, however, a small set of times when it would be nice to be able to omit arguments in the stub configuration, because they truly do not matter for the subject under test.

Would you be open to adding an option to allow extra arguments to be ignored?

```ts
const myDouble = jest.fn()

// potential APIs
// new method
when(myDouble).partiallyCalledWith("foo").mockReturnValue("bar")
// options object in when
when(myDouble, { ignoreExtraArgs: true }).calledWith("foo").mockReturnValue("bar")

assert(myDouble("foo") === "bar")
assert(myDouble("foo", "other", "stuff") === "bar")
```

## risks and alternatives

- Alternative: just use jest's asymmetric matchers
    - An `expect.anything()` or several could do the trick in most cases, at the expense of verbosity
- Alternative: write your own `mockImplementation` for these use cases
- Alternative: write a little `when` and `calledWith` wrapper and deal with the problem in userland
- Risk: this would be a powerful feature that could be easily misused
    - I really misused Sinon's partial matching in my early days of learning how to mock
    - The default of this library is to require all arguments for a reason
- Risk: maintenance burden

## prior art

[testdouble.js](https://github.com/testdouble/testdouble.js), which has a very similar API to `jest-when` and seems to me to come from the same stubbing lineage, has an [`ignoreExtraArgs`](https://github.com/testdouble/testdouble.js/blob/master/docs/5-stubbing-results.md#ignoreextraargs) option during stub configuration.

[sinon.js](https://sinonjs.org/) does partial stubbing by default in its [`calledWith`](https://sinonjs.org/releases/v10.0.1/spies/#spycalledwitharg1-arg2-) assertion. I think this is a bad default, but folks coming from Sinon would at least be familiar with this behavior.

(Edit) after a deeper issue search that I should've completed before hitting submit, I ran into #36. I think this request is related, but distinct, because I'm talking specifically about rare cases where trailing arguments have no meaning to the subject nor stub.

## actual use case

In this particular case, we're mocking out React function components. We have a collaborator component, and we want to verify that some child component is properly arranged with the correct props:

```ts
import * as React from 'react'
import { when } from 'jest-when'
import { render, screen } from '@testing-library/react'
import { SomeChild } from './SomeChild'  // export const SomeChild = (props) => (<>{...}</>)
import { SomeParent } from './SomeParent'

jest.mock('./SomeChild')

describe('SomeParent component', () => {
  it('should pass xyz prop down to SomeChild', () => {
    when(ShomeChild).calledWith({ xyz: '123' }).mockReturnValue(<span>hello world</span>)

    render(<SomeParent xyz="123" />)
    expect(screen.getByText('hello world'))
  })
})
```

You (or rather, I) would expect this to work! However, **it does not**. For legacy reasons that have nothing to do with the code under test, the React rendering system passes two arguments the the `SomeChild` function, `props`, and something else called `refOrContext`. All of our function components are written to accept a single function parameter, `props`, so the fact that React is passing in a second parameter doesn't do anything because it's always ignored.

We can fix the test by doing...

```ts
when(ShomeChild)
  .calledWith({ xyz: '123' }, expect.anything())
  .mockReturnValue(<span>hello world</span>)
```

...but it's a little more verbose and, more importantly to me, muddies how the test communicates its intent to the reader

### PR DESCRIPTION
**Pull request by timkindberg**
- New when.allArgs matcher. Instructs jest-when to hand over the full set of arguments to the matcher. This allows a dev to perform more complex matches, or in some cases write more succinct matching logic.

Closes #66 

@mcous Can you please review and see if this might meet your needs? I have one test in particular that address your specific use case.

Still needs:
- [x] Documentation

### CODE REVIEW

File: src/when.test.js [hunk 1]
```diff
@@ -232,6 +232,78 @@ describe('When', () => {
...
+          .mockReturnValue('hello world')
+
+        const propsOf = propsToMatch => when.allArgs(([props, refOrContext], equals) => equals(props, propsToMatch))
+
+        when(SomeChild)
+          .calledWith(propsOf({ xyz: '123' }))
+          .mockReturnValue('hello world')
+
+        expect(SomeChild({ xyz: '123' })).toEqual('hello world')
+      })
```
Comment (timkindberg) [inline src/when.test.js:276]: @mcous ^^

Comment (mcous) [reply src/when.test.js:276]: I think this `propsOf` utility is how we'd end up actually using this, so this test case looks solid to me

File: src/when.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ const checkArgumentMatchers = (expectCall, args) => (match, matcher, i) => {
   }
 
   if (isFunctionMatcher) {
-    return matcher(arg)
+    return matcher(arg, utils.equals)
```
Comment (mcous) [inline src/when.js:36]: Is this new functionality for the regular function matchers, too? Seems nice to have, and (if I'm not reading this incorrectly) might be worth covering with a test-case in the regular `when(matcher)` context

**Review by mcous — APPROVED** [review]

File: src/when.js (hunk 2 above)
Comment (timkindberg) [reply src/when.js:36]: good point, yeah I just realized folks might like the power of that equals util in their own custom function matchers. I'll add a test.

**Comment by timkindberg** [comment]
@mcous I'm really liking this feature! Check out the documentation changes I made.

File: README.md [hunk 3]
```diff
@@ -211,6 +211,56 @@ expect(fn(4)).toEqual('way!')
...
+expect(fn(3, 666)).toEqual('yay!')
+expect(fn(-100, 2, 3.234234, 234, 90e3)).toEqual('yay!')
+expect(fn(123, 'not a number')).toBeUndefined()
+```
+
+E.g. Single arg match:
+```javascript
+const argAtIndex = (index, matcher) => (args, equals) => equals(args[index], matcher)
+
+when(fn).calledWith(argAtIndex(0, expect.any(Number))).mockReturnValue('yay!')
```
Comment (mcous) [inline README.md:241]: Is this example missing a `when.allArgs` somewhere?

**Review by mcous — COMMENTED** [review]
Docs look great! And yeah, I like that this solution combines generality with a little bit of an extra hoop to jump through to prevent misuse or overuse

File: README.md (hunk 3 above)
Comment (timkindberg) [reply README.md:241]: good catch! fixed.

**Pull request merged**